nixos/modules/services/matrix/mautrix-whatsapp.nix

```diff
@@ -0,0 +1,199 @@
+{
+  lib,
+  config,
+  pkgs,
+  ...
+}: let
+  cfg = config.services.mautrix-whatsapp;
+  dataDir = "/var/lib/mautrix-whatsapp";
+  registrationFile = "${dataDir}/whatsapp-registration.yaml";
+  settingsFile = "${dataDir}/config.json";
+  settingsFileUnsubstituted = settingsFormat.generate "mautrix-whatsapp-config-unsubstituted.json" cfg.settings;
+  settingsFormat = pkgs.formats.json {};
+  appservicePort = 29318;
+in {
+  imports = [];
+  options.services.mautrix-whatsapp = {
+    enable = lib.mkEnableOption "mautrix-whatsapp, a puppeting/relaybot bridge between Matrix and WhatsApp.";
+
+    settings = lib.mkOption {
+      type = settingsFormat.type;
+      default = {};
+      description = lib.mdDoc ''
+        {file}`config.yaml` configuration as a Nix attribute set.
+        Configuration options should match those described in
+        [example-config.yaml](https://github.com/mautrix/whatsapp/blob/master/example-config.yaml).
+        Secret tokens should be specified using {option}`environmentFile`
+        instead of this world-readable attribute set.
+      '';
+      example = {
+        appservice = {
+          database = {
+            type = "postgres";
+            uri = "postgresql:///mautrix_whatsapp?host=/run/postgresql";
+          };
+          id = "whatsapp";
+          ephemeral_events = false;
+        };
+        bridge = {
+          history_sync = {
+            request_full_sync = true;
+          };
+          private_chat_portal_meta = true;
+          mute_bridging = true;
+          encryption = {
+            allow = true;
+            default = true;
+            require = true;
+          };
+          provisioning = {
+            shared_secret = "disable";
+          };
+          permissions = {
+            "example.com" = "user";
+          };
+        };
+      };
+    };
+    environmentFile = lib.mkOption {
+      type = lib.types.nullOr lib.types.path;
+      default = null;
+      description = lib.mdDoc ''
+        File containing environment variables to be passed to the mautrix-whatsapp service,
+        in which secret tokens can be specified securely by optionally defining a value for
+        `MAUTRIX_WHATSAPP_BRIDGE_LOGIN_SHARED_SECRET`.
+      '';
+    };
+
+    serviceDependencies = lib.mkOption {
+      type = with lib.types; listOf str;
+      default = lib.optional config.services.matrix-synapse.enable "matrix-synapse.service";
+      defaultText = lib.literalExpression ''
+        optional config.services.matrix-synapse.enable "matrix-synapse.service"
+      '';
+      description = lib.mdDoc ''
+        List of Systemd services to require and wait for when starting the application service.
+      '';
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    services.mautrix-whatsapp.settings = {
```

Comment:
Can't these be set in the default section above? (which is done in the mautrix-telegram module)

Reply:
I encontered an [error with nix docs](https://github.com/NixOS/nixpkgs/actions/runs/5743104020/job/15566635874) when referencing another module from the default attribute set.
That's when I started doing some research on that topic. I found [this section in nixos manual](https://nixos.org/manual/nixos/stable/#sec-settings-nix-representable) and following that approach moving the default values to the config attribute set.

Reply:
> If you insist, I could moving only the homeserver.domain attribute to settings and leave the rest in `default`.

This sounds like a better solution, at least it reduces the amount of `lib.mkDefault`

Reply:
Done.
But shouldn't the manual be updated then also?
I read Example 32 from the [NixOS manual](https://nixos.org/manual/nixos/stable/#sec-settings-nix-representable) as default values should be defined using `lib.mkDefault`.

Reply:
AFIAK these two approaches should be equivalent in action. Just personally prefer the better looking one.

Reply:
Thanks for the clarification! :)